Statistics not working with Group Replication / MySQL 8 - cannot use Stroom Status Dashoard

After the fix for #1462  (for running STROOM v7 against MySQL 8.0) in **stroom_core_test-v7.0-beta.36**, STROOM now runs, and can be logged into, but there is a problem with statictics.

The defauls Stroom Status Dashboard cannot be used properly: on opening it generated multiple popups like `Unable to locate datasource 946a7f0f-a59a-11e6-bcd4-0242ac110002`, with about 5 different UUIDS, and also `No data source has been chosen to search`.

In the stroom log we see a numeber of `FlywayException` lines like:
```ERROR  [2020-06-03T12:51:07.582Z] [Stroom Job - ScheduledExecutor] stroom.db.util.AbstractFlyWayDbModule -  Error migrating stroom-statistics database
org.flywaydb.core.internal.command.DbMigrate$FlywayMigrateException:
Migration V07_00_00_001__Schema.sql failed
------------------------------------------
SQL State  : HY000
Error Code : 3098
Message    : The table does not comply with the requirements by an external plugin.
Location   : stroom/statistics/impl/sql/db/migration/V07_00_00_001__Schema.sql (/stroom/file:/stroom/stroom-app-all.jar!/stroom/statistics/impl/sql/db/migration/V07_00_00_001__Schema.sql)
Line       : 121
Statement  : CALL add_col_to_stat_val_src()

        at org.flywaydb.core.internal.command.DbMigrate.doMigrateGroup(DbMigrate.java:370)
        at org.flywaydb.core.internal.command.DbMigrate.access$200(DbMigrate.java:54)
        at org.flywaydb.core.internal.command.DbMigrate$3.call(DbMigrate.java:284)
        at org.flywaydb.core.internal.jdbc.TransactionTemplate.execute(TransactionTemplate.java:74)
        at org.flywaydb.core.internal.command.DbMigrate.applyMigrations(DbMigrate.java:281)
        at org.flywaydb.core.internal.command.DbMigrate.migrateGroup(DbMigrate.java:246)
        at org.flywaydb.core.internal.command.DbMigrate.access$100(DbMigrate.java:54)
        at org.flywaydb.core.internal.command.DbMigrate$2.call(DbMigrate.java:164)
        at org.flywaydb.core.internal.command.DbMigrate$2.call(DbMigrate.java:161)
        at org.flywaydb.core.internal.database.mysql.MySQLNamedLockTemplate.execute(MySQLNamedLockTemplate.java:60)
        at org.flywaydb.core.internal.database.mysql.MySQLConnection.lock(MySQLConnection.java:147)
        at org.flywaydb.core.internal.schemahistory.JdbcTableSchemaHistory.lock(JdbcTableSchemaHistory.java:155)
        at org.flywaydb.core.internal.command.DbMigrate.migrateAll(DbMigrate.java:161)
        at org.flywaydb.core.internal.command.DbMigrate.migrate(DbMigrate.java:139)
        at org.flywaydb.core.Flyway$1.execute(Flyway.java:1395)
        at org.flywaydb.core.Flyway$1.execute(Flyway.java:1356)
        at org.flywaydb.core.Flyway.execute(Flyway.java:1711)
        at org.flywaydb.core.Flyway.migrate(Flyway.java:1356)
        at stroom.db.util.AbstractFlyWayDbModule.performMigration(AbstractFlyWayDbModule.java:48)
        at stroom.db.util.AbstractDataSourceProviderModule.getConnectionProvider(AbstractDataSourceProviderModule.java:59)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.base/java.lang.reflect.Method.invoke(Unknown Source)
        at com.google.inject.internal.ProviderMethod$ReflectionProviderMethod.doProvision(ProviderMethod.java:286)
        at com.google.inject.internal.ProviderMethod.doProvision(ProviderMethod.java:173)
        at com.google.inject.internal.InternalProviderInstanceBindingImpl$CyclicFactory.provision(InternalProviderInstanceBindingImpl.java:185)
        at com.google.inject.internal.InternalProviderInstanceBindingImpl$CyclicFactory.get(InternalProviderInstanceBindingImpl.java:162)
        at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
        at com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:168)
        at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:39)
        at com.google.inject.internal.SingleParameterInjector.inject(SingleParameterInjector.java:42)
        at com.google.inject.internal.SingleParameterInjector.getAll(SingleParameterInjector.java:65)
        at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:113)
        at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:91)
        at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:306)
        at com.google.inject.internal.FactoryProxy.get(FactoryProxy.java:62)
        at com.google.inject.internal.SingleParameterInjector.inject(SingleParameterInjector.java:42)
        at com.google.inject.internal.SingleParameterInjector.getAll(SingleParameterInjector.java:65)
        at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:113)
        at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:91)
        at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:306)
        at com.google.inject.internal.SingleParameterInjector.inject(SingleParameterInjector.java:42)
        at com.google.inject.internal.SingleParameterInjector.getAll(SingleParameterInjector.java:65)
        at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:113)
        at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:91)
        at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:306)
        at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
        at com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:168)
        at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:39)
        at com.google.inject.internal.SingleParameterInjector.inject(SingleParameterInjector.java:42)
        at com.google.inject.internal.SingleParameterInjector.getAll(SingleParameterInjector.java:65)
        at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:113)
        at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:91)
        at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:306)
        at com.google.inject.internal.FactoryProxy.get(FactoryProxy.java:62)
        at com.google.inject.internal.InjectorImpl$1.get(InjectorImpl.java:1050)
        at stroom.job.impl.ScheduledTaskExecutor.create(ScheduledTaskExecutor.java:194)
        at stroom.job.impl.ScheduledTaskExecutor.execute(ScheduledTaskExecutor.java:135)
        at stroom.job.impl.ScheduledTaskExecutor.lambda$startup$0(ScheduledTaskExecutor.java:89)
        at stroom.security.impl.SecurityContextImpl.asProcessingUser(SecurityContextImpl.java:282)
        at stroom.job.impl.ScheduledTaskExecutor.lambda$startup$1(ScheduledTaskExecutor.java:87)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at java.base/java.util.concurrent.FutureTask.runAndReset(Unknown Source)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.flywaydb.core.internal.sqlscript.FlywaySqlScriptException:
Migration V07_00_00_001__Schema.sql failed
------------------------------------------
SQL State  : HY000
Error Code : 3098
Message    : The table does not comply with the requirements by an external plugin.
Location   : stroom/statistics/impl/sql/db/migration/V07_00_00_001__Schema.sql (/stroom/file:/stroom/stroom-app-all.jar!/stroom/statistics/impl/sql/db/migration/V07_00_00_001__Schema.sql)
Line       : 121
Statement  : CALL add_col_to_stat_val_src()

        at org.flywaydb.core.internal.sqlscript.DefaultSqlScriptExecutor.handleException(DefaultSqlScriptExecutor.java:253)
        at org.flywaydb.core.internal.sqlscript.DefaultSqlScriptExecutor.executeStatement(DefaultSqlScriptExecutor.java:202)
        at org.flywaydb.core.internal.sqlscript.DefaultSqlScriptExecutor.execute(DefaultSqlScriptExecutor.java:125)
        at org.flywaydb.core.internal.resolver.sql.SqlMigrationExecutor.execute(SqlMigrationExecutor.java:77)
        at org.flywaydb.core.internal.command.DbMigrate.doMigrateGroup(DbMigrate.java:367)
        ... 67 common frames omitted
Caused by: java.sql.SQLException: The table does not comply with the requirements by an external plugin.
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:129)
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97)
        at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
        at com.mysql.cj.jdbc.StatementImpl.executeInternal(StatementImpl.java:782)
        at com.mysql.cj.jdbc.StatementImpl.execute(StatementImpl.java:666)
        at com.zaxxer.hikari.pool.ProxyStatement.execute(ProxyStatement.java:95)
        at com.zaxxer.hikari.pool.HikariProxyStatement.execute(HikariProxyStatement.java)
        at org.flywaydb.core.internal.jdbc.JdbcTemplate.executeStatement(JdbcTemplate.java:235)
        at org.flywaydb.core.internal.sqlscript.StandardSqlStatement.execute(StandardSqlStatement.java:42)
        at org.flywaydb.core.internal.sqlscript.DefaultSqlScriptExecutor.executeStatement(DefaultSqlScriptExecutor.java:189)
        ... 70 common frames omitted
ERROR  [2020-06-03T12:51:07.591Z] [Stroom Job - ScheduledExecutor] stroom.job.impl.ScheduledTaskExecutor -  Unable to provision, see the following errors:

1) Error in custom provider, org.flywaydb.core.internal.command.DbMigrate$FlywayMigrateException:
Migration V07_00_00_001__Schema.sql failed
------------------------------------------
SQL State  : HY000
Error Code : 3098
Message    : The table does not comply with the requirements by an external plugin.
Location   : stroom/statistics/impl/sql/db/migration/V07_00_00_001__Schema.sql (/stroom/file:/stroom/stroom-app-all.jar!/stroom/statistics/impl/sql/db/migration/V07_00_00_001__Schema.sql)
Line       : 121
Statement  : CALL add_col_to_stat_val_src()

  at stroom.db.util.AbstractDataSourceProviderModule.getConnectionProvider(Unknown Source) (via modules: stroom.app.guice.AppModule -> stroom.app.guice.CoreModule -> stroom.statistics.impl.sql.SQLStatisticsModule)
  while locating stroom.statistics.impl.sql.SQLStatisticsDbConnProvider
    for the 1st parameter of stroom.statistics.impl.sql.search.StatisticsSearchServiceImpl.<init>(Unknown Source)
  while locating stroom.statistics.impl.sql.search.StatisticsSearchServiceImpl
  while locating stroom.statistics.impl.sql.search.StatisticsSearchService
    for the 2nd parameter of stroom.statistics.impl.sql.search.SqlStatisticStoreFactory.<init>(Unknown Source)
  while locating stroom.statistics.impl.sql.search.SqlStatisticStoreFactory
    for the 3rd parameter of stroom.statistics.impl.sql.search.SqlStatisticsSearchResponseCreatorManager.<init>(Unknown Source)
  at stroom.statistics.impl.sql.search.SqlStatisticsSearchResponseCreatorManager.class(Unknown Source)
  while locating stroom.statistics.impl.sql.search.SqlStatisticsSearchResponseCreatorManager
    for the 1st parameter of stroom.statistics.impl.sql.search.SQLStatisticSearchModule$EvictExpiredElements.<init>(Unknown Source)
  while locating stroom.statistics.impl.sql.search.SQLStatisticSearchModule$EvictExpiredElements
  while locating java.lang.Runnable annotated with @com.google.inject.internal.Element(setName=,uniqueId=672, type=MAPBINDER, keyType=stroom.job.api.ScheduledJob)

1 error
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
 
+* Issue **#1632** : SQL stats now compatible with MySQL 8 Group Replication
+
 * Issue **#1650** : Use consitent blue colour.
 
 * Issue **#1627** : Fix Up/Down buttons on Rule Set screen. Now keeps selection after use.

--- a/stroom-statistics/stroom-statistics-impl-sql/src/main/resources/stroom/statistics/impl/sql/db/migration/V07_00_00_001__Schema.sql
+++ b/stroom-statistics/stroom-statistics-impl-sql/src/main/resources/stroom/statistics/impl/sql/db/migration/V07_00_00_001__Schema.sql
@@ -40,12 +40,12 @@ CREATE TABLE IF NOT EXISTS SQL_STAT_KEY (
 -- Table structure for table sql_stat_val (idempotent)
 --
 CREATE TABLE IF NOT EXISTS SQL_STAT_VAL (
-  TIME_MS				bigint(20) NOT NULL,
-  PRES					tinyint(4) NOT NULL,
-  VAL_TP 				tinyint(4) NOT NULL,
-  VAL					double     NOT NULL,
-  CT					bigint(20) NOT NULL,
-  FK_SQL_STAT_KEY_ID	bigint(20) NOT NULL,
+  TIME_MS               bigint(20) NOT NULL,
+  PRES                  tinyint(4) NOT NULL,
+  VAL_TP                tinyint(4) NOT NULL,
+  VAL                   double     NOT NULL,
+  CT                    bigint(20) NOT NULL,
+  FK_SQL_STAT_KEY_ID    bigint(20) NOT NULL,
   PRIMARY KEY (FK_SQL_STAT_KEY_ID, TIME_MS, VAL_TP, PRES),
   CONSTRAINT 			SQL_STAT_VAL_FK_STAT_KEY_ID
       FOREIGN KEY (FK_SQL_STAT_KEY_ID)
@@ -94,6 +94,7 @@ ALTER TABLE SQL_STAT_VAL_SRC MODIFY COLUMN PROCESSING tinyint(1) NOT NULL DEFAUL
 -- Copy data into the job node table
 --
 DROP PROCEDURE IF EXISTS add_col_to_stat_val_src;
+DROP PROCEDURE IF EXISTS add_id_to_stat_val_src;
 DELIMITER //
 CREATE PROCEDURE add_col_to_stat_val_src ()
 BEGIN
@@ -119,8 +120,80 @@ BEGIN
     MODIFY COLUMN CT BIGINT(20) NOT NULL;
 
 END//
+CREATE PROCEDURE add_id_to_stat_val_src ()
+BEGIN
+    -- Idempotent
+    IF NOT EXISTS (
+            SELECT NULL
+            FROM INFORMATION_SCHEMA.COLUMNS
+            WHERE TABLE_NAME = 'SQL_STAT_VAL_SRC'
+            AND COLUMN_NAME = 'ID') THEN
+        SELECT CONCAT('Beginning migration of ',
+                          ' table SQL_STAT_VAL_SRC in database ',
+                          database(), ' to version 7 (adding id field to stats)'
+                          '.');
+        -- We are missing the ID (autoincrement, pk) col.
+        -- Create a new table and copy values back
+        RENAME TABLE SQL_STAT_VAL_SRC TO SQL_STAT_VAL_SRC_OLD;
+
+        -- Create the table with the new structure
+        CREATE TABLE IF NOT EXISTS SQL_STAT_VAL_SRC (
+          ID bigint(20) auto_increment,
+          TIME_MS bigint(20) NOT NULL,
+          NAME varchar(766) NOT NULL,
+          VAL_TP tinyint(4) NOT NULL,
+          VAL double NOT NULL,
+          CT bigint(20) NOT NULL,
+          PROCESSING bit(1) NOT NULL DEFAULT 0,
+          PRIMARY KEY (ID)
+        ) ENGINE=InnoDB DEFAULT CHARSET=latin1;
+
+        -- Copy the data over
+        INSERT INTO SQL_STAT_VAL_SRC (
+          TIME_MS, NAME, VAL_TP, VAL, CT, PROCESSING
+        ) SELECT TIME_MS, NAME, VAL_TP, VAL, CT, PROCESSING FROM SQL_STAT_VAL_SRC_OLD;
+
+        -- Check that the correct number of columns have been copied
+        BEGIN
+            DECLARE row_count_before integer;
+            DECLARE row_count_after integer;
+
+            SELECT COUNT(*)
+              INTO row_count_before
+              FROM SQL_STAT_VAL_SRC_OLD;
+
+            SELECT COUNT(*)
+              INTO row_count_after
+              FROM SQL_STAT_VAL_SRC;
+
+            SELECT CONCAT('There are ', row_count_before,
+                          ' records to migrate.');
+            -- Delete the old table if copy was successful
+            IF row_count_before = row_count_after THEN
+              DROP TABLE SQL_STAT_VAL_SRC_OLD;
+              SELECT CONCAT('Successfully migrated ',row_count_before,
+                            ' records into table SQL_STAT_VAL_SRC in database ',
+                            database(), ' to version 7 (added id field to stats)',
+                            '.');
+            ELSE
+            -- Attempt rollback
+              DROP TABLE SQL_STAT_VAL_SRC;
+              RENAME TABLE SQL_STAT_VAL_SRC_OLD TO SQL_STAT_VAL_SRC;
+
+              SELECT CONCAT('Failed to copy all records! Copied ',
+                          row_count_after,' of ',
+                          row_count_before,
+                          ' records from table SQL_STAT_VAL_SRC while migrating ',
+                          database(), ' to version 7 (adding id field to stats)',
+                          '.');
+            END IF;
+         END;
+    END IF;
+END//
 DELIMITER ;
 CALL add_col_to_stat_val_src();
+CALL add_id_to_stat_val_src();
 DROP PROCEDURE add_col_to_stat_val_src;
+DROP PROCEDURE add_id_to_stat_val_src;
 
 SET SQL_NOTES=@OLD_SQL_NOTES;

--- a/stroom-statistics/stroom-statistics-impl-sql/src/main/resources/stroom/statistics/impl/sql/db/migration/V07_00_00_001__Schema.sql
+++ b/stroom-statistics/stroom-statistics-impl-sql/src/main/resources/stroom/statistics/impl/sql/db/migration/V07_00_00_001__Schema.sql
@@ -67,12 +67,14 @@ MODIFY COLUMN VAL double;
 -- Table structure for table sql_stat_val_src (idempotent)
 --
 CREATE TABLE IF NOT EXISTS SQL_STAT_VAL_SRC (
+  ID 				bigint(20)   auto_increment,
   TIME_MS			bigint(20)   NOT NULL,
   NAME 				varchar(766) NOT NULL,
   VAL_TP 			tinyint(4)   NOT NULL,
   VAL				double       NOT NULL,
   CT				bigint(20)   NOT NULL,
-  PROCESSING        bit(1)       NOT NULL DEFAULT 0
+  PROCESSING        bit(1)       NOT NULL DEFAULT 0,
+  PRIMARY KEY (ID)
 ) ENGINE=InnoDB DEFAULT CHARSET=latin1;
 
 -- used for deletions of old data where the FK is not involved

